### PR TITLE
Implement `transform_descendants()`, rewrite `jj parallelize`

### PR DESCRIPTION
**Pull request by martinvonz**
@bnjmnt4n: This PR includes the `transform_descendants()` function.

@emesterhazy: FYI, this is how I'm planning to rewrite `jj parallelize`.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [inline cli/tests/test_parallelize_command.rs:253]: I'm not sure if I understand why this is an ok result since 3 is not parallel with respect to 1 after the command runs.

I think I do understand how this happens though. Since 2 is not part of the target set, it retains 1 as a parent. For the same reason 3 retains 2 as a parent. And so there are no changes. I guess this is a result that seems surprising at first, but makes sense and is totally correct if you understand how the command *really* works.

Maybe you can add a comment to the test with the explanation above?

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [inline cli/tests/test_parallelize_command.rs:245]: nit: the test name should change since this is no longer a failure case per-se.

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [inline cli/tests/test_parallelize_command.rs:436]: nit: It doesn't really matter, but is there any way to preserve the graph order better? I haven't looked at exactly how `jj log` orders the commits, but I'm guessing modification time plays a role?

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [inline cli/tests/test_parallelize_command.rs:512]: Very cool! nit: remove "failure" from the test name.

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [inline cli/tests/test_parallelize_command.rs:554]: nit: remove "failure" from the test name.

**Review by emesterhazy — COMMENTED** [review]
I left some comments. Overall this is really cool and the descendants rebasing API seems very powerful!

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:436]: It's not exactly the modification time, or at least not the time recorded in the commit. It's the order they were written to the index. There's not much we can do here without a lot of ugly hacks (like removing these commits from the index and then adding them back).

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:245]: will do. this is just a draft to show how it will work. i also plan to add library-level tests of `transform_descendants()`.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:253]: Right. I should update the help text to clarify that the command effectively says that the inputs are independent among themselves, but that external dependencies are preserved.

File: cli/tests/test_parallelize_command.rs
Comment (emesterhazy) [reply cli/tests/test_parallelize_command.rs:436]: Yeah, not worth it then.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:253]: Done.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:245]: Done.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:512]: Done.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:554]: Done

**Review by martinvonz — COMMENTED** [review]
This is now ready for review

File: lib/src/repo.rs
Comment (yuja) [inline lib/src/repo.rs:1180]: I assumed `update_rewritten_references()` is idempotent (so `transform_descendants()` can be called multiple times without clearing the `parent_mapping`), but the assumption might not hold if the working-copy commit got abandoned? Just guess.

File: lib/src/rewrite.rs
Comment (yuja) [inline lib/src/rewrite.rs:161]: nit: perhaps, this can be rewritten as `.iter().position()` + `.splice()`, but we might have to ensure that `new_parents` have no dupes in `self.new_parents`

File: cli/src/commands/parallelize.rs
Comment (yuja) [inline cli/src/commands/parallelize.rs:84]: nit: `.extend_from_slice()`

File: cli/src/commands/parallelize.rs
Comment (yuja) [inline cli/src/commands/parallelize.rs:113]: nit: redundant `.cloned()`

File: cli/src/commands/parallelize.rs
Comment (yuja) [inline cli/src/commands/parallelize.rs:125]: nit: `.iter().cloned()`

File: cli/src/commands/parallelize.rs
Comment (yuja) [inline cli/src/commands/parallelize.rs:131]: nit: might be better to add rewriter method that looks up "rewritten" new parents and set them, but I'm not sure.

File: cli/tests/test_parallelize_command.rs
Comment (yuja) [inline cli/tests/test_parallelize_command.rs:333]: nit: maybe better to use `get_log_output_with_parents()` globally? It's uneasy to read the ancestry from the graph.

File: cli/src/commands/parallelize.rs
Comment (martinvonz) [reply cli/src/commands/parallelize.rs]: Good idea. Done.

File: cli/tests/test_parallelize_command.rs
Comment (martinvonz) [reply cli/tests/test_parallelize_command.rs:333]: Done in a commit just before this commit.

File: lib/src/repo.rs
Comment (martinvonz) [reply lib/src/repo.rs:1180]: I have also thought it would be idempotent. What's the case you're thinking of where it might not be? If the working copy pointed to an abandoned commit, it would be moved somewhere else (typically the parents) and would no longer point to a commit in `parent_mapping`, so a subsequent `update_rewritten_references()` should be a no-op. Maybe you're thinking of a more complex case?

Comment (yuja) [reply lib/src/repo.rs:1180]: > Maybe you're thinking of a more complex case?

No, it's just a guess that there might be a problem because `update_wc_commits()` does non-trivial thing. Good to know that the call is idempotent.

**Comment by martinvonz** [comment]
@yuja: This PR still requires an approval. I don't mean to rush you, just making sure you didn't forget it (I sometimes think that I had approved a PR that I had not approved). Take your time if you need.

**Review by yuja — APPROVED** [review]
I just thought it might be better to be reviewed by others, but the change looks good to me.

**Comment by martinvonz** [comment]
> I just thought it might be better to be reviewed by others, but the change looks good to me.

Makes sense. Thanks for reviewing!

@bnjmnt4n, @emesterhazy, and/or @ilyagr might want to take a look. I'll merge it based on Yuya's review and to unblock e.g. Benjamin's `rebase --insert-before/-after` series. I'm happy to send fixes as separate PRs if you have any thoughts.

**Pull request merged**
